bindings/azure/signalr/signalr.go

```diff
@@ -0,0 +1,194 @@
+)
+
+const (
+	errorPrefix         = "azure signalr error:"
+	logPrefix           = "azure signalr:"
+	connectionStringKey = "connectionString"
+	hubKey              = "hub"
+	groupKey            = "group"
+	userKey             = "user"
+)
+
+// NewSignalR creates a new pub/sub based on Azure SignalR
+func NewSignalR() *SignalR {
+	return &SignalR{
+		tokens:     make(map[string]signalrCachedToken),
+		httpClient: &http.Client{Timeout: 30 * time.Second},
+	}
+}
+
+type signalrCachedToken struct {
+	token      string
+	expiration time.Time
+}
+
+// SignalR is an output binding for Azure SignalR
+type SignalR struct {
+	endpoint   string
+	accessKey  string
+	version    string
+	hub        string
+	tokens     map[string]signalrCachedToken
+	httpClient *http.Client
+}
+
+// Init is responsible for initializing the SignalR output based on the metadata
+func (s *SignalR) Init(metadata bindings.Metadata) error {
+	connectionString, ok := metadata.Properties[connectionStringKey]
+	if !ok || connectionString == "" {
+		return fmt.Errorf("missing connection string")
+	}
+
+	if hub, ok := metadata.Properties[hubKey]; ok && hub != "" {
+		s.hub = hub
+	}
+
+	// Expected: Endpoint=https://<servicename>.service.signalr.net;AccessKey=<access key>;Version=1.0;
+	connectionValues := strings.Split(strings.TrimSpace(connectionString), ";")
+	for _, connectionValue := range connectionValues {
+		if i := strings.Index(connectionValue, "="); i != -1 && len(connectionValue) > (i+1) {
+			k := connectionValue[0:i]
+			switch k {
+			case "Endpoint":
+				s.endpoint = connectionValue[i+1:]
+				if s.endpoint[len(s.endpoint)-1] == '/' {
+					s.endpoint = s.endpoint[:len(s.endpoint)-1]
+				}
+			case "AccessKey":
+				s.accessKey = connectionValue[i+1:]
+			case "Version":
+				s.version = connectionValue[i+1:]
+			}
+		}
+	}
+
+	if s.endpoint == "" {
+		return fmt.Errorf("missing endpoint in connection string")
+	}
+
+	if s.accessKey == "" {
+		return fmt.Errorf("missing access key in connection string")
+	}
+
+	return nil
+}
+
+func (s *SignalR) resolveAPIURL(req *bindings.WriteRequest) (string, error) {
+	hub := s.hub
+	if hub == "" {
+		hubFromRequest, ok := req.Metadata[hubKey]
+		if !ok || hubFromRequest == "" {
+			return "", fmt.Errorf("%s missing hub", errorPrefix)
+		}
+
+		hub = hubFromRequest
+	}
+
+	var url string
+	if group, ok := req.Metadata[groupKey]; ok && group != "" {
+		url = fmt.Sprintf("%s/api/v1/hubs/%s/groups/%s", s.endpoint, hub, group)
+	} else if user, ok := req.Metadata[userKey]; ok && user != "" {
+		url = fmt.Sprintf("%s/api/v1/hubs/%s/users/%s", s.endpoint, hub, user)
+	} else {
+		url = fmt.Sprintf("%s/api/v1/hubs/%s", s.endpoint, hub)
+	}
+
+	return url, nil
+}
+
+func (s *SignalR) sendMessageToSignalR(url string, token string, data []byte) error {
+	httpReq, err := http.NewRequest("POST", url, bytes.NewBuffer(data))
+	if err != nil {
+		return err
+	}
+
+	httpReq.Header.Add("Authorization", "Bearer "+token)
```

Comment:
Set vs Add?

Reply:
In this case, it doesn't matter.

Here are the differences:

```
Set sets the header entries associated with key to the single element value. It replaces any existing values associated with key. The key is case insensitive; it is canonicalized by textproto.CanonicalMIMEHeaderKey. To use non-canonical keys, assign to the map directly.
```

```
Add adds the key, value pair to the header. It appends to any existing values associated with key. The key is case insensitive; it is canonicalized by CanonicalHeaderKey.
```

Reply:
Let’s call set for both

Reply:
I missed the Set call. Should have consistency, so either set or add for both.